examples/pytorch/rgcn/experimental/entity_classify_dist.py

```diff
@@ -0,0 +1,536 @@
+        return h
+
+def init_emb(shape, dtype):
+    arr = th.zeros(shape, dtype=dtype)
+    nn.init.uniform_(arr, -1.0, 1.0)
+    return arr
+
+class DistEmbedLayer(nn.Module):
+    r"""Embedding layer for featureless heterograph.
+    Parameters
+    ----------
+    dev_id : int
+        Device to run the layer.
+    g : DistGraph
+        training graph
+    num_of_ntype : int
+        Number of node types
+    embed_size : int
+        Output embed size
+    sparse_emb: bool
+        Whether to use sparse embedding
+        Default: False
+    dgl_sparse_emb: bool
+        Whether to use DGL sparse embedding
+        Default: False
+    embed_name : str, optional
+        Embed name
+    """
+    def __init__(self,
+                 dev_id,
+                 g,
+                 num_of_ntype,
+                 embed_size,
+                 sparse_emb=False,
+                 dgl_sparse_emb=False,
+                 embed_name='node_emb'):
+        super(DistEmbedLayer, self).__init__()
+        self.dev_id = dev_id
+        self.num_of_ntype = num_of_ntype
+        self.embed_size = embed_size
+        self.embed_name = embed_name
+        self.sparse_emb = sparse_emb
+
+        if sparse_emb:
+            if dgl_sparse_emb:
+                self.node_embeds = dgl.distributed.DistEmbedding(g.number_of_nodes(),
+                                                                 self.embed_size,
+                                                                 embed_name,
+                                                                 init_emb)
+            else:
+                self.node_embeds = th.nn.Embedding(g.number_of_nodes(), self.embed_size, sparse=self.sparse_emb)
+                nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
+        else:
+            self.node_embeds = th.nn.Embedding(g.number_of_nodes(), self.embed_size)
+            nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
+
+    def forward(self, node_ids, node_tids, features):
+        """Forward computation
+        Parameters
+        ----------
+        node_ids : tensor
+            node ids to generate embedding for.
+        node_ids : tensor
+            node type ids
+        features : list of features
+            list of initial features for nodes belong to different node type.
+            If None, the corresponding features is an one-hot encoding feature,
+            else use the features directly as input feature and matmul a
+            projection matrix.
+        Returns
+        -------
+        tensor
+            embeddings as the input of the next layer
+        """
+        embeds = th.empty(node_ids.shape[0], self.embed_size)
+        for ntype in range(self.num_of_ntype):
+            assert features[ntype] is None, 'Currently Dist RGCN only support non input feature'
+            loc = node_tids == ntype
+            embeds[loc] = self.node_embeds(node_ids[loc])
```

Comment:
this is differentiable? if the underlying is index_copy, this seems very slow.

Reply:
When we introduce input features later, we have to do this.
This is not slow as the shape of loc and node_tids is not very large.